src/Symfony/Component/Serializer/Normalizer/ObjectNormalizer.php

```diff
@@ -110,15 +112,20 @@ protected function extractAttributes($object, $format = null, array $context = a
      */
     protected function getAttributeValue($object, $attribute, $format = null, array $context = array())
     {
-        if (null !== $this->classDiscriminatorResolver) {
-            $mapping = $this->classDiscriminatorResolver->getMappingForMappedObject($object);
+        $cacheKey = get_class($object).'::'.$attribute;
 
-            if (null !== $mapping && $attribute == $mapping->getTypeProperty()) {
-                return $this->classDiscriminatorResolver->getTypeForMappedObject($object);
+        if (!isset(self::$isDiscriminatorAttributeCache[$cacheKey])) {
+            self::$isDiscriminatorAttributeCache[$cacheKey] = false;
+            if (null !== $this->classDiscriminatorResolver) {
+                $mapping = $this->classDiscriminatorResolver->getMappingForMappedObject($object);
+
+                if (null !== $mapping && $attribute == $mapping->getTypeProperty()) {
+                    self::$isDiscriminatorAttributeCache[$cacheKey] = true;
+                }
             }
         }
 
-        return $this->propertyAccessor->getValue($object, $attribute);
+        return self::$isDiscriminatorAttributeCache[$cacheKey] ? $this->classDiscriminatorResolver->getTypeForMappedObject($object) : $this->propertyAccessor->getValue($object, $attribute);
```

Comment:
We would even improve the performance more: for a given object, regardless of the property, the result of `getTypeForMappedObject` will always be the same.
We could use `SplObjectStorage` to call this method only one time for every object.

Reply:
So we would have two caches, one to know if the current property we are getting  the value is the discriminator field of the class and an other one to get the value of this discriminator mapping.

Does the later worth it? It would probably be used only when the same object is present more than once in the serialized objects graph.

Reply:
`getAttributeValue` is called for every attribute of a given class, isn't it?

Reply:
Got it!

Reply:
I pushed a v2 https://github.com/symfony/symfony/pull/28889/commits/0bd95ec5a970e570d5ebfc28dbf30cf4715ac25d and the improvement is less significant: https://blackfire.io/profiles/compare/fd66194c-e3a7-4bc1-bea4-c67d2677148f/graph

IMHO we need two levels of cache, one to known if a class has a discriminator, and one to get the property name (roughly what is done in v2).

WDYT?